bin.src/generateCommands.py

```diff
@@ -0,0 +1,581 @@
+    def getSubclass(typeName:str) -> type:
+        return ScienceStep.__subclasses[typeName]
+
+    @staticmethod
+    def registerAsSubclass(typeName:str):
+        def helper(cls):
+            ScienceStep.__subclasses[typeName] = cls
+            cls.typeName = typeName
+            return cls
+        return helper
+
+    # To use (subclass instance).execute(),
+    # remember to override `commandName`
+    commandName = None
+
+    def execute(self, visits, dataDir, calib, rerun, processes=1, verbose=False):
+        visits = visitsToString(visits)
+
+        command = [
+            self.commandName,
+            dataDir,
+            f"--calib={os.path.join(dataDir, calib)}",
+            f"--rerun={rerun}",
+            "--id", f"visit={visits}",
+            "-j", processes,
+            ]
+
+        if self.configs:
+            command += ["--config"] + self.configs
+
+        print(f"{shellCommand(command)} || return 1")
+
+
+@ScienceStep.registerAsSubclass("reduceExposure")
+class ReduceExposureStep(ScienceStep):
+    commandName = "reduceExposure.py"
+
+@ScienceStep.registerAsSubclass("mergeArms")
+class MergeArmsStep(ScienceStep):
+    commandName = "mergeArms.py"
+
+@ScienceStep.registerAsSubclass("calculateReferenceFlux")
+class CalculateReferenceFluxStep(ScienceStep):
+    commandName = "calculateReferenceFlux.py"
+
+@ScienceStep.registerAsSubclass("fluxCalibrate")
+class FluxCalibrateStep(ScienceStep):
+    commandName = "fluxCalibrate.py"
+
+@ScienceStep.registerAsSubclass("coaddSpectra")
+class CoaddSpectraStep(ScienceStep):
+    commandName = "coaddSpectra.py"
+
+class ScienceBlock:
+    steps = [] # will be replaced in processYaml()
+
+    def __init__(self, blockName:str, yamlBlock:Mapping[str,Any]):
+        self.name = blockName
+        self.visits = _getVisits(yamlBlock)
+
+        self.pipeline = {}
+        for step in self.steps:
+            block = yamlBlock.get("pipeline", {}).get(step, {})
+            self.pipeline[step] = ScienceStep.getSubclass(step).fromYaml(block)
+
+    def execute(self, dataDir, calib, rerun, steps=None, processes=1, verbose=False):
+        """Execute a science block"""
+
+        if steps is None:
+            steps = self.steps
+
+        if verbose:
+            print(f"Processing science block '{self.name}'", file=sys.stderr)
+
+        for step in self.steps:
+            self.pipeline[step].execute(self.visits, dataDir, calib, f"{rerun}/pipeline", processes=processes, verbose=verbose)
+
+
+def _getConfigs(yamlBlock:Mapping[str,Any]) -> List[str]:
+    """
+    Extract config: ... and configs: ... from yamlBlock.
+    """
+    configs = []
+
+    if "config" in yamlBlock:
+        configs.append(yamlBlock["config"])
+
+    if "configs" in yamlBlock:
+        for c in yamlBlock["configs"]:
+            configs.append(c)
+
+    return configs
+
+def _getVisits(yamlBlock:Mapping[str,Any]) -> List[int]:
+    """
+    Extract visit: ... and visits: ... from yamlBlock.
```

Comment:
This is also potentially confusing syntax.

Reply:
`visit` and `visits` have been merged to `visit.